src/par_iter/chain.rs

```diff
@@ -0,0 +1,173 @@
+        let a = self.a.drive(left);
+        let b = self.b.drive(right);
+        reducer.reduce(a, b)
+    }
+}
+
+impl<A, B> ExactParallelIterator for ChainIter<A, B>
+    where A: ExactParallelIterator, B: ExactParallelIterator<Item=A::Item>
+{
+    fn len(&mut self) -> usize {
+        self.a.len() + self.b.len()
+    }
+}
+
+impl<A, B> IndexedParallelIterator for ChainIter<A, B>
+    where A: IndexedParallelIterator, B: IndexedParallelIterator<Item=A::Item>
+{
+    fn with_producer<CB>(mut self, callback: CB) -> CB::Output
+        where CB: ProducerCallback<Self::Item>
+    {
+        let a_len = self.a.len();
+        return self.a.with_producer(CallbackA {
+            callback: callback,
+            a_len: a_len,
+            b: self.b,
+        });
+
+        struct CallbackA<CB, B> {
+            callback: CB,
+            a_len: usize,
+            b: B,
+        }
+
+        impl<CB, B> ProducerCallback<B::Item> for CallbackA<CB, B>
+            where B: IndexedParallelIterator,
+                  CB: ProducerCallback<B::Item>,
+        {
+            type Output = CB::Output;
+
+            fn callback<A>(self, a_producer: A) -> Self::Output
+                where A: Producer<Item=B::Item>
+            {
+                return self.b.with_producer(CallbackB {
+                    callback: self.callback,
+                    a_len: self.a_len,
+                    a_producer: a_producer,
+                });
+            }
+        }
+
+        struct CallbackB<CB, A> {
+            callback: CB,
+            a_len: usize,
+            a_producer: A,
+        }
+
+        impl<CB, A> ProducerCallback<A::Item> for CallbackB<CB, A>
+            where A: Producer,
+                  CB: ProducerCallback<A::Item>,
+        {
+            type Output = CB::Output;
+
+            fn callback<B>(self, b_producer: B) -> Self::Output
+                where B: Producer<Item=A::Item>
+            {
+                let producer = ChainProducer::new(self.a_len,
+                                                  self.a_producer,
+                                                  b_producer);
+                self.callback.callback(producer)
+            }
+        }
+
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////
+
+pub struct ChainProducer<A, B>
+    where A: Producer, B: Producer<Item=A::Item>
+{
+    a_len: usize,
+    a: A,
+    b: B,
+}
+
+impl<A, B> ChainProducer<A, B>
+    where A: Producer, B: Producer<Item=A::Item>
+{
+    fn new(a_len: usize, a: A, b: B) -> Self {
+        ChainProducer { a_len: a_len, a: a, b: b }
+    }
+}
+
+impl<A, B> Producer for ChainProducer<A, B>
+    where A: Producer, B: Producer<Item=A::Item>
+{
+    fn cost(&mut self, len: usize) -> f64 {
+        let a_len = min(self.a_len, len);
+        let b_len = len - a_len;
+        self.a.cost(a_len) + self.b.cost(b_len)
+    }
+
+    fn split_at(self, index: usize) -> (Self, Self) {
+        if index <= self.a_len {
```

Comment:
TBH I'm not thrilled with that, since I have to keep splitting at the boundary to get producers that I know will be empty.  I also tried wrapping them in `Option`, but that made it a lot more difficult to deal with.  So this is a little weird, but it keeps things simple.